dynatrace/src/main/java/com/dynatrace/opentelemetry/metric/CumulativeToDeltaConverter.java

```diff
@@ -0,0 +1,153 @@
+      this.cache.invalidate(identifier);
+      return newValue;
+    }
+    final CacheValue cacheValue = this.cache.getIfPresent(identifier);
+
+    Double deltaValue;
+    final long pointTimestamp =
+        (point.getEpochNanos() > 0)
+            ? TimeUnit.NANOSECONDS.toMillis(point.getEpochNanos())
+            : System.currentTimeMillis();
+    if (cacheValue == null) {
+      // no value in the cache yet or already expired
+      deltaValue = null;
+    } else {
+      if (cacheValue.getTimeMillis() > pointTimestamp) {
+        // the current point is older than the one in the cache, so leave the previous value.
+        // the delta between a value and itself is 0.
+        return 0d;
+      } else {
+        // point is newer than the stored data, so calculate delta.
+        deltaValue = newValue - cacheValue.getNumber().doubleValue();
+      }
+    }
+
+    // update the cache
+    this.cache.put(identifier, new CacheValue(newValue, pointTimestamp));
+    return deltaValue;
+  }
+
+  /**
+   * Convert a total counter to a delta. The identity of a counter is defined by its name and its
+   * dimensions.
+   *
+   * @param metricName The name of the metric.
+   * @param point The {@link LongPointData} object containing the data point.
+   * @return The delta to the previous value if there is one with the same identifier in the cache
+   *     or the value itself if there is not.
+   */
+  public Long convertLongTotalToDelta(String metricName, LongPointData point) {
+    String identifier = createIdentifier(metricName, point.getLabels(), "LONG");
+
+    long newValue = point.getValue();
+    final CacheValue cacheValue = this.cache.getIfPresent(identifier);
+
+    Long deltaValue;
+    final long pointTimestamp =
+        (point.getEpochNanos() > 0)
+            ? TimeUnit.NANOSECONDS.toMillis(point.getEpochNanos())
+            : System.currentTimeMillis();
+    if (cacheValue == null) {
+      // no value in the cache yet or already expired
+      deltaValue = null;
+    } else {
+      if (cacheValue.getTimeMillis() > pointTimestamp) {
+        // the current point is older than the one in the cache, so leave the previous value.
+        // the delta between a value and itself is 0.
+        return 0L;
+      } else {
+        // point is newer than the stored data, so calculate delta.
+        deltaValue = newValue - cacheValue.getNumber().longValue();
+      }
+    }
+
+    // update the cache
+    this.cache.put(identifier, new CacheValue(newValue, pointTimestamp));
+    return deltaValue;
+  }
+
+  // VisibleForTesting
+  void reset() {
+    this.cache.invalidateAll();
+  }
+
+  private static String getSortedLabelsString(Labels labels) {
+    if (labels.isEmpty()) {
+      return "";
+    }
+    List<AbstractMap.SimpleEntry<String, String>> keyValuePairs = new ArrayList<>(labels.size());
+    labels.forEach((k, v) -> keyValuePairs.add(new AbstractMap.SimpleEntry<>(k, v)));
```

Comment:
I guess we can't sort this `Labels` first? Because if we could, then the second for loop would not be necessary.

Reply:
The `Labels` class does not have a sort method on it, but in the default implementation labels are sorted automatically. I haven't found any guarantee for that however, and its also possible to overwrite the labels class and change the default behavior. Therefore, I would advocate for keeping the sort logic, as it should not need to sort anything in the default case.

Reply:
Makes sense!